organizer check #298

minsooerickim opened the pull request:
currently working on getting query and resolver to work

minsooerickim committed:
query added

minsooerickim committed:
unrecognized resolver fix

minsooerickim committed:
in progress of using queryRef to do conditional rendering for the sideNav

minsooerickim committed:
in progress of using queryRef to do conditional rendering for the sideNav

minsooerickim committed:
removed isOrganizer query and just added as a field under preexisting userSideNavQuery

minsooerickim committed:
redirect to dashboard implemented

minsooerickim committed:
download csv file from gcloud and parse for email

minsooerickim committed:
download csv file from gcloud and parse for email

DevelopmentByDavid commented on app/client/src/features/accounts/UserSideNav.tsx:
```diff
@@ -88,11 +90,21 @@ export function UserSideNavLoader() {
     usePreloadedQuery(USER_SIDE_NAV_QUERY, queryRef);
     const [selected, setSelected] = React.useState<Keys | undefined>(findTab(router.pathname));
 
+    const { displaySnack } = useSnack();
+    React.useEffect(() => {
+        if (!data.isOrganizer && router.pathname === '/organizations/me') {
+            router.push('/dashboard');
+            displaySnack('Organizers-only path!');
+        }
+    }, [data.isOrganizer, displaySnack, router]);
+
```
I think this would be better delegated to a component.  Take a look at https://github.com/c2c-project/prytaneum/blob/dev/app/client/src/components/ConditionalRender.tsx

Maybe you could make a version of this that's for the user type? 

```tsx
<RoleGuard organizer>
  <SomeOrganizerOnlyComponent />
</RoleGuard>
<RoleGuard organizer moderator>
   // An organizer or a moderator could access inside of here
  <SomeComponent />
</RoleGuard>
<Roleguard organizer redirect="/path/to/go/to/if/not/an/organizer">
  <SomethingSensitive />
</RoleGuard>
```

DevelopmentByDavid commented on app/client/src/layout/SideNav/SideNav.tsx:
```diff
@@ -27,7 +27,7 @@ export function PreloadUserSideNav(props: PreloadedUserSideNavProps) {
     if (!queryRef) return <Loader />;
 
-    return <UserSideNav queryRef={queryRef} {...props} />;
+    return <UserSideNav queryRef={queryRef} {...props}/>;
```
Space got erased.  Double check that prettier is working properly.

DevelopmentByDavid commented on app/server/package.json:
```diff
@@ -28,6 +29,7 @@
         "fastify-cors": "^6.0.2",
         "fastify-multipart": "^5.3.1",
         "form-data": "^4.0.0",
+        "fs": "^0.0.1-security",
```
Hmm, why did fs get added here, that doesn't seem correct?  fs should already be part of nodejs itself.  If you check, this package doesn't actually do anything. https://www.npmjs.com/package/fs

Be careful when installing packages, you don't want to accidentally install a virus :smile:

DevelopmentByDavid replied:
fs should be removed.

minsooerickim replied:
removed!

DevelopmentByDavid commented on app/server/src/features/organizations/resolvers.ts:
```diff
@@ -10,6 +12,38 @@ const toUserId = toGlobalId('User');
+            }
+            if (process.env.NODE_ENV === 'development') return true;
+            const email = await Organization.findEmailByViewerId(ctx.viewer.id, ctx.prisma);
+            const bucketName = 'organizers';
+            const fileName = 'list-of-organizers.csv';
+            const storage = new Storage({
+                projectId: process.env.GCP_PROJECT_ID,
+                keyFilename: process.env.GOOGLE_APPLICATION_CREDENTIALS,
+            });
+            const options = {
+                destination: './downloadedFile.txt',
+            };
+            const file = storage.bucket(bucketName).file(fileName);
+            let found = false;
+            try {
+                await file.download(options);
+            } catch (err) {
+                throw new Error(err);
+            }
+            const allFileContents = fs.readFileSync('./downloadedFile.txt', 'utf-8');
+            allFileContents.split(/\r?\n?,/).forEach((line) => {
+                if (line === email?.email) {
+                    found = true;
+                }
+            });
+            return found;
```
I think a good chunk of this logic could be delegated to another Organizer method.  Lets say `Organizer.isOnOrganizerList`

minsooerickim replied:
done!

DevelopmentByDavid commented on app/server/src/features/organizations/resolvers.ts:
```diff
@@ -10,6 +12,38 @@ const toUserId = toGlobalId('User');
+                keyFilename: process.env.GOOGLE_APPLICATION_CREDENTIALS,
+            });
+            const options = {
+                destination: './downloadedFile.txt',
```
I think the name of the file should probably be an environment variable, say `ORGANIZER_FILE_PATH`.

DevelopmentByDavid commented on package.json:
```diff
@@ -42,6 +42,8 @@
         "typescript": "4.3.2"
     },
     "dependencies": {
+        "@google-cloud/storage": "^5.19.2",
+        "babel-eslint": "^10.1.0",
```
I think it was probably a mistake to add these to the root package.json, yeah?  Make sure to remove them if so.

DevelopmentByDavid replied:
These should probably get removed, yeah?

minsooerickim replied:
yea that must have been a mistake. They are removed now!

DevelopmentByDavid reviewed (changes requested):
Great progress!

minsooerickim committed:
style(server): creates isOnOrganizerList method
a whole chuk of connecting to google cloud and checking for organizer is done in a new method

minsooerickim committed:
in progress of making RoleGuard component

minsooerickim committed:
in progress of making RoleGuard component

DevelopmentByDavid commented on app/client/src/components/RoleGuard.tsx:
```diff
@@ -0,0 +1,14 @@
+import router from 'next/router';
+
+export interface RoleGuardProps {
+    organizer?: boolean;
+    moderator?: boolean;
```
Could you add some simple comments here? For example, `true if the moderator is an organizer`.  React props kinda breaks best practice re: naming booleans. Normally, you'd want `isOrganizer`, but with React props booleans are typicallly just a noun. 

In summary, what you have is correct, please just add some basic comments.

DevelopmentByDavid commented on app/client/src/components/RoleGuard.tsx:
```diff
@@ -0,0 +1,14 @@
+}
+
+export function RoleGuard({ organizer = false, moderator = false, children }: RoleGuardProps){
+    if (organizer) return <>{children}</>
+    else router.push('/dashboard');
+    if (moderator) return <>{children}</>
```
This could be simplified.

```jsx
if (organizer || moderator) return <>{children}</>
return <></>
```

Although, it may be more complicated depending on what you wanna do per my comment further up in this file.

DevelopmentByDavid commented on app/client/src/components/RoleGuard.tsx:
```diff
@@ -0,0 +1,14 @@
+    children: JSX.Element | JSX.Element[];
+}
+
+export function RoleGuard({ organizer = false, moderator = false, children }: RoleGuardProps){
```
Could you add a comment on what this component is for?  For example, `By default this component will not render any children unless the user is a moderator or an organizer of a particular event.`

Arguably, this component that should be using `useEvent` and checking whether or not the user is a moderator or an organizer.  Then, you could conditionally render children based on the parameters passed in by the caller. This would then be a more robust component.  What do you think?

Examples:
```tsx
// organizer only
<RoleGuard organizer>
...
</RoleGuard>

// moderator only (organizers have access because they are already a moderator)
<RoleGuard moderator>
...
</RoleGuard>
```

DevelopmentByDavid commented on app/client/src/features/accounts/UserSideNav.tsx:
```diff
@@ -9,15 +9,18 @@ import AssignmentIcon from '@mui/icons-material/Assignment';
 import { AnimateSharedLayout /* motion */ } from 'framer-motion';
 import { useRouter } from 'next/router';
 import { graphql, usePreloadedQuery, PreloadedQuery } from 'react-relay';
-
```
Please separate package imports and local imports by a newline.

DevelopmentByDavid commented on app/client/src/components/RoleGuard.tsx:
```diff
@@ -0,0 +1,14 @@
+import router from 'next/router';
+
+export interface RoleGuardProps {
```
Maybe add a `redirect` string param so that can be customized?  If it's missing, then we don't redirect at all.  Some things we may just not display to users, but overall the page stays the same regardless of moderator vs user.

DevelopmentByDavid commented on app/client/src/features/accounts/UserSideNav.tsx:
```diff
@@ -88,11 +92,22 @@ export function UserSideNavLoader() {
+    React.useEffect(() => {
+        if (!data.isOrganizer && router.pathname === '/organizations/me') {
+            router.push('/dashboard');
+            displaySnack('Organizers-only path!');
```
I would say "not allowed." We would want to leak unnecessary permission policies.

minsooerickim replied:
With the new role guard feature, we don't need this code at all so I commented it all out for now, but I still did change the string to "not allowed"!

DevelopmentByDavid commented on app/client/src/features/accounts/UserSideNav.tsx:
```diff
@@ -126,14 +141,38 @@ export function UserSideNav({ queryRef, onClick }: UserSideNavProps) {
-                    </StyledListItemIcon>
-                    <ListItemText primary='My Organizations' />
-                </StyledListItem>
+                <RoleGuard organizer={data.isOrganizer}>
+                    <>
+                        <StyledSubheader>Organizations</StyledSubheader>
+                        <StyledDivider />
+
+                        <StyledListItem
+                            onClick={handleClick('My Organizations')}
+                            selected={selected === 'My Organizations'}
+                        >
+                            <StyledListItemIcon>
+                                <ListIcon />
+                            </StyledListItemIcon>
+                            <ListItemText primary='My Organizations' />
+                        </StyledListItem>
+                    </>
+                </RoleGuard>
```
In the current implementation of the `RoleGuard` component, I think this would redirect if the user is not an organizer, which I'm not sure is the behavior we want.  Is that correct?

minsooerickim replied:
yea that was very unfinished code from what I remember, but I did push more changes that should make more sense now!

DevelopmentByDavid commented on app/server/src/features/organizations/methods.ts:
```diff
@@ -9,6 +9,44 @@ import {
+/** 
+ * given a user id, return the associated email
+*/
+export async function findEmailByViewerId(userId: string, prisma: PrismaClient) {
```
`findEmailByUserId`

minsooerickim committed:
fix(packages): removed fs from package.json

minsooerickim committed:
fix(spacing): separated package imports and local imports by a newline

minsooerickim committed:
fix(typo): changed findEmailByViewerId to findEmailByUserId

minsooerickim committed:
feat(roleguard): added error comments

minsooerickim committed:
fix(typo): fix typo

minsooerickim committed:
feat(RoleGuard): delegate RoleGuard.tsx for just isOrganizer

minsooerickim committed:
feat(RoleGuard): moved isOrganizer resolver+helper method under accounts from organization

minsooerickim committed:
feat(RoleGuard): add isOrganizer under in useUserFragment

minsooerickim committed:
fix(RoleGuard): minor fix for RoleGuard

minsooerickim committed:
fix(RoleGuard): minor fix for RoleGuard

minsooerickim commented:
The reason why `isOrganizer` isn't returning correctly at the moment is because the `isOrganizer` resolver isn't being called anywhere in the code. Right now, I am just fetching the `isOrganizer` field of the `User` which by default is null I believe (which is why the children aren't being returned correctly at the moment);

I'm not sure how to go about updating the `isOrganizer` field everytime we fetch for its value; I appreciate any feedback, thank you!

kadey001 committed:
Server Unit + Integration Testing Setup (#313)

* chore(server): added testing libraries and set up jest

* test(server): set up tests for account methods

* chore(server): added testing libraries and scripts

* chore(server): export redis instance and updated node env checks

* chore(env): added env test file

* test(server): implemented basic integration tests

* chore(eslint): removed unused patterns

* test(server): updated methods test file and refactored

* chore(server): updated integration test paths

* chore(server): added redis vars to env files and env check

* ci(db): added test db script

* ci(server): added prisma db push script and updated test ci script

* ci(workflow): added server testing

* chore(server mocks): added prisma docs comment for prisma mock

* chore(jest): updated scripts and added coverage path ignore patterns

* chore(server): removed faker

* chore(server): added comments and prisma deleteMany cleanup

* chore(server): formatting and more account unit tests

* test(account methods): fixed null password test

* test(unit): aaa formatting and comments

* chore(prisma mock): added comment about default export

* test(integration): aaa formatting and comments

* chore(server): wrapped tests in desc block and comments cleanup

* test(integration): added usr organizations tests

* fix(db): fixed remove-orphans warning with flag

* chore(server): updated comment and test wording

* feat(db): added prisma db seeding

kadey001 committed:
User Guides (#319)

* feat(client): set up and added user guides

* feat(guides): made demo components for getting started guide

* feat(guides): added moderator guide

* fix(pages): updated guide function names

* chore(guides): updated titles

* feat(guides): added participant guide

* fix(guides): added padding to headings to offset appbar

* chore(guides): removed old outdated guide

kadey001 committed:
chore(server): removed unused packages

kadey001 committed:
chore(organizer): isOrganizer prisma check and added mutation to add and remove permission

kadey001 committed:
fix(roleguard): fixed rolegaurd check and action

kadey001 committed:
fix(client): added missing images back

kadey001 committed:
fix(client): updated redirects, isOrganizer query, and added auth roleguard

kadey001 committed:
fix(e2e): updated expected redirect

kadey001 committed:
chore(prettier): formatting

kadey001 committed:
chore(role guard): simplified, added comment, and removed unessiary query

kadey001 committed:
chore(logout): removed page reload after logout

kadey001 committed:
chore(client): removed unused images

kadey001 committed:
chore(server): restore gcloud pubsub version

kadey001 committed:
chore(packagse): removed unused packages

kadey001 committed:
fix(client): event form now displays correctly

kadey001 committed:
fix(client): added refresh for logout on live page

kadey001 committed:
chore(yarn): ran yarn install

DevelopmentByDavid commented:
Going to close and reopen this PR. With all the force pushes and temporary deploys it's really hard to keep track of what's what. We'll get this merged though.

DevelopmentByDavid commented:
Also, this PR seems to have a lot of scope creep. I'm breaking it up into multiple most likely.
